source/common/common/utility.h

```diff
@@ -288,4 +289,54 @@ class RegexUtil {
                                std::regex::flag_type flags = std::regex::optimize);
 };
 
+/**
+ * Maintains sets of numeric intervals. As new intervals are added, existing ones in the
+ * set are combined so that no overlapping intervals remain in the representation.
+ *
+ * Value can be any type that is comparable with <.
+ */
+template <typename Value> class IntervalSet {
+public:
+  typedef std::pair<Value, Value> Interval;
+
+  // Inserts a new interval into the set, merging any overlaps.
+  void insert(Value left, Value right) {
+    auto left_pos = intervals_.lower_bound(Interval(left, left));
```

Comment:
Might be worthwhile to toss in an `ASSERT(!(right < left))` at the beginning of this method.

Reply:
Done, and actually the code was not handling an empty inclusion of [x, x) correctly.  Fixed & added testcase for that, as well as an assert.